### PR TITLE
Add support for cpu_get_num_physical_cores() on Windows

### PR DESCRIPTION
**Pull request by Septa2112**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

Add a function `cpu_get_num_cores_win(bool print_physical_core_num)` to detect the physical and logical cores on Windows. Now the program will successfully detect all sockets and CPUs and set the default number of threads more accurately on Windows.

## Test Platform
- OS
  - Windows 10
  - Windows 11
- CPU
  - Intel Xeon CPU E5-2699 v4 (22 physical core per socket,  **2 Sockets in total**)
  - Intel i7-1185G7 (4 physical cores)
  - Intel Core Ultra5 125H (4 P-core + 8 E-core + 2 LPE-core)

## Before
When I run `llama-cli.exe -m <my_model> -p <prompt>`, the output is like this:

### Intel Xeon CPU E5-2699 v4 (2 Sockets)
```shell
system_info: n_threads = 22 / 44 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | 
```
### Intel i7-1185G7
```shell
system_info: n_threads = 4 / 8 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | 
```
### Intel MTL Ultra5
```shell
system_info: n_threads = 9 / 18 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | 
```

### Intel Xeon CPU E5-2699 v4 (2 Sockets)
```shell
system_info: n_threads = 44 / 88 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | 
```
### Intel i7-1185G7
```shell
system_info: n_threads = 4 / 8 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | 
```
### Intel MTL Ultra5
```shell
system_info: n_threads = 14 / 18 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | 
```

### CODE REVIEW

**Comment by Septa2112** [comment]
@mofosyne Could you have a look at my PR? Or anyone others? Thanks.

File: common/common.h [hunk 1]
```diff
@@ -44,7 +44,9 @@ struct llama_control_vector_load_info;
 //
 // CPU utils
 //
-
+#ifdef _WIN32
+int32_t cpu_get_num_cores_win(bool print_physical_core_num);
+#endif
```
Comment (ggerganov) [inline common/common.h:49]: Instead of declaring this new function, just inline the windows code in the existing `cpu_get_num_physical_cores`

**Review by ggerganov — COMMENTED** [review]
We will need some reports on Windows that this change works as expected

File: common/common.h (hunk 1 above)
Comment (Septa2112) [reply common/common.h]: Yes, I have deleted the new function in my latest commit.

**Comment by Septa2112** [comment]
I noticed that the CI `CI / windows-latest-cmake (msvc-arm64, -G "Ninja Multi-Config"...` failed, indicating that the Windows API I used was not detected. 
However, another CI about arm64 `CI / windows-latest-cmake (llvm-arm64, -G "Ninja Multi-Config" `is successful(tested myself using GitHub Actions). 

I noticed that these two CIs were added due to #7191

To address this, I attempted to skip the ARM64 platform with the following condition: 
```
#elif defined(_WIN32) && (!defined(__ARM_NEON) || !defined(_M_ARM) || !defined(_M_ARM64) || !defined(__aarch64__) || !defined(__arm__) || !defined(__ARM_ARCH))
```

However, despite my repeated attempts, none of these macros related to the ARM architecture were detected in this CI environment.



I suspect there might be an issue with the toolchain, but unfortunately, I don't have any ARM64+Windows machine for testing. Could you kindly suggest any solutions or approaches to resolve this issue?

**Comment by Septa2112** [comment]
I carefully reviewed the CI error and found that the issue seems to be with `MinGW64+ARM64` on windows. Therefore, I made the following modification
```
#elif defined(_WIN32) && (_WIN32_WINNT >= 0x0601) && !defined(__MINGW64__)
```
to bypass the issue with the toolchain on ARM64+Windows by detecting __MINGW64__.

**Review by ggerganov — APPROVED** [review]

**Comment by hmartinez82** [comment]
@Septa2112 Once this is merged I'll take a look at why ARM64 MinGW64 is not building :-)

**Comment by Septa2112** [comment]
@hmartinez82 That's great. Let's see when the PR will be merged. Perhaps need more test on windows.

**Comment by hmartinez82** [comment]
@Septa2112 Just as I suspected, after glancing over the PR, there's nothing there that would not work. 
I just tried building for ARM64 with both x-compiling from a x64 MSVC (17.10) and natively with MSYS2's Clang in a CLANGARM64 environment.

The build was successful in both cases.

**Comment by Septa2112** [comment]
@hmartinez82 Thanks for you testing. But if I remove this macro, the corresponding CI checks will fail. Maybe something else is causing the issue?

This CI was introduced in the issue #7191.

**Comment by hmartinez82** [comment]
Ahh ok, I thought that the issue was during build time. I'll try to replicate what the CI does. Thank you.

**Pull request merged**